internal/provider/data_source_account.go

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package provider
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/hashicorp/boundary/api"
+	"github.com/hashicorp/boundary/api/accounts"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+func dataSourceAccount() *schema.Resource {
+	return &schema.Resource{
+		Description: "The boundary_account data source allows you to find a Boundary account.",
+		ReadContext: dataSourceAccountRead,
+
+		Schema: map[string]*schema.Schema{
+			NameKey: {
+				Description:  "The name of the account to retrieve.",
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			AuthMethodIdKey: {
+				Description:  "The auth method ID that will be queried for the account.",
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			IDKey: {
+				Description: "The ID of the retrieved account.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			DescriptionKey: {
+				Description: "The description of the retrieved account.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			TypeKey: {
+				Description: "The type of the account",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			ScopeKey: {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						IDKey: {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						NameKey: {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						TypeKey: {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						DescriptionKey: {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						ParentScopeIdKey: {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceAccountRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	md := meta.(*metaData)
+
+	name := d.Get(NameKey).(string)
+	authMethodId := d.Get(AuthMethodIdKey).(string)
+
+	acl := accounts.NewClient(md.client)
+	accountsList, err := acl.List(ctx, authMethodId,
+		accounts.WithFilter(FilterWithItemNameMatches(name)),
+	)
+	accounts := accountsList.GetItems()
+	if err != nil {
+		return diag.Errorf("error calling list account: %v", err)
+	}
+	if accounts == nil {
+		return diag.Errorf("no accounts found")
+	}
+	if len(accounts) == 0 {
+		return diag.Errorf("no matching account found: %v", err)
+	}
+
+	if len(accounts) > 1 {
+		return diag.Errorf("error found more than 1 account: %v", err)
+	}
+
+	var accountId string
+	for _, account := range accounts {
```

Comment:
If we have the line above that always checks to make sure we only have 1 account (`if len(accounts) > 1`), do we still need the for loop? Won't there always be only 1 account in the list at this point?

Reply:
Good point. What do you suggest here? 

We could just get rid of that block and go directly to
```
arr, err := acl.Read(ctx, accounts[0].Id)
```

Reply:
For Boundary List calls where we can have multiple items, I wonder if we should also always return 1 time for datasources or a list which could have either 1 item in it or more than 1 item? Right now we follow a pattern of always returning 1 item for a data source.

Would customers come back to ask for a new datasource that can return all items of a resource based on attributes/filters passed in because they do not want to repeat calls to a datasource to get more data.

I'll leave it to Louis to share his thoughts on this.

cc: @louisruch

Reply:
Good question. I had similar thoughts. Right now, the current data sources just do a search by `name`. Is that all we want here? Do we want to filter by other attributes? 

Alternatively, we could just start with this current approach for now and iterate in the future.

Reply:
As far as I am aware a datasource must always match to a single resource Boundary. This is needed so you can bring in a resource by some filter and then use the id or other values in other resources. And this will be done via tf mapping.

As for future filters I think we definitely will want to use more than just name, but its a great place to start and we can expand as needed.